File: go/v2/payment_methods_test.go
package v2_test

import (
	"context"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"

	tesote "github.com/tesote/sdk/go"
	"github.com/tesote/sdk/go/v2"
)

const pmJSON = `{"id":"p1","method_type":"bank_account","currency":"VES","label":null,"details":{},"verified":false,"verified_at":null,"last_used_at":null,"counterparty":null,"tesote_account":null,"created_at":"x","updated_at":"x"}`

func TestPaymentMethods_List_VerifiedFilter(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Query().Get("verified") != "true" {
			t.Errorf("verified = %q", r.URL.Query().Get("verified"))
		}
		_, _ = io.WriteString(w, `{"items":[`+pmJSON+`],"has_more":false,"limit":50,"offset":0}`)
	}))
	defer srv.Close()
	c := v2.New(newClient(t, srv))
	verified := true
	out, err := c.PaymentMethods.List(context.Background(), v2.PaymentMethodsListOptions{Verified: &verified})
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if len(out.Items) != 1 {
		t.Errorf("items = %+v", out)
	}
}

func TestPaymentMethods_Get(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = io.WriteString(w, pmJSON)
	}))
	defer srv.Close()
	c := v2.New(newClient(t, srv))
	out, err := c.PaymentMethods.Get(context.Background(), "p1")
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if out.ID != "p1" || out.MethodType != "bank_account" {
		t.Errorf("pm = %+v", out)
	}
}

func TestPaymentMethods_Get_NotFound(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		_, _ = io.WriteString(w, `{"error":"x","error_code":"PAYMENT_METHOD_NOT_FOUND"}`)
	}))
	defer srv.Close()
	c := v2.New(newClient(t, srv))
	_, err := c.PaymentMethods.Get(context.Background(), "missing")
	if !errors.Is(err, tesote.ErrPaymentMethodNotFound) {
		t.Errorf("err = %v", err)
	}
}

func TestPaymentMethods_Create_ValidationError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
		_, _ = io.WriteString(w, `{"error":"bad","error_code":"VALIDATION_ERROR"}`)
	}))
	defer srv.Close()
	c := v2.New(newClient(t, srv))
	_, err := c.PaymentMethods.Create(context.Background(), v2.PaymentMethodMutateOptions{
		Input: v2.PaymentMethodInput{MethodType: "bank_account"},
	})
	var typed *tesote.ValidationError
	if !errors.As(err, &typed) {
		t.Errorf("err = %T %v", err, err)
	}
}

func TestPaymentMethods_Update_PATCH(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPatch {
			t.Errorf("method = %s", r.Method)
		}
		_, _ = io.WriteString(w, pmJSON)
	}))
	defer srv.Close()
	c := v2.New(newClient(t, srv))
	label := "primary"
	_, err := c.PaymentMethods.Update(context.Background(), "p1", v2.PaymentMethodMutateOptions{
		Input: v2.PaymentMethodInput{Label: &label},
	})
	if err != nil {
		t.Fatalf("Update: %v", err)
	}
}

func TestPaymentMethods_Delete_204(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete {
			t.Errorf("method = %s", r.Method)
		}
		w.WriteHeader(http.StatusNoContent)
	}))
	defer srv.Close()
	c := v2.New(newClient(t, srv))
	if err := c.PaymentMethods.Delete(context.Background(), "p1", v2.PaymentMethodDeleteOptions{}); err != nil {
		t.Fatalf("Delete: %v", err)
	}
}

func TestPaymentMethods_Delete_InUseConflict(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusConflict)
		_, _ = io.WriteString(w, `{"error":"in use","error_code":"VALIDATION_ERROR"}`)
	}))
	defer srv.Close()
	c := v2.New(newClient(t, srv))
	err := c.PaymentMethods.Delete(context.Background(), "p1", v2.PaymentMethodDeleteOptions{})
	var typed *tesote.ValidationError
	if !errors.As(err, &typed) {
		t.Errorf("err = %T %v", err, err)
	}
}

File: packages/php/tests/V2/PaymentMethodsTest.php
<?php

declare(strict_types=1);

namespace Tesote\Sdk\Tests\V2;

use Tesote\Sdk\Errors\PaymentMethodNotFoundException;
use Tesote\Sdk\Errors\ValidationException;
use Tesote\Sdk\Http\CurlResult;
use Tesote\Sdk\Models\PaymentMethod;
use Tesote\Sdk\Models\PaymentMethodList;
use Tesote\Sdk\Tests\Support\TestCaseBase;
use Tesote\Sdk\V2\Client;

final class PaymentMethodsTest extends TestCaseBase
{
    /**
     * @return array<string, mixed>
     */
    private static function fixture(): array
    {
        return [
            'id' => 'pm-1',
            'method_type' => 'bank_account',
            'currency' => 'VES',
            'label' => 'Main',
            'details' => ['bank_code' => '0102', 'account_number' => '****1234', 'holder_name' => 'ACME'],
            'verified' => true,
            'verified_at' => '2026-04-28T00:00:00Z',
            'last_used_at' => null,
            'counterparty' => ['id' => 'cp-1', 'name' => 'Vendor'],
            'tesote_account' => null,
            'created_at' => '2026-01-01T00:00:00Z',
            'updated_at' => '2026-04-01T00:00:00Z',
        ];
    }

    public function testList(): void
    {
        $this->enqueueOk([
            'items' => [self::fixture()],
            'has_more' => false,
            'limit' => 50,
            'offset' => 0,
        ]);
        $client = new Client(['transport' => $this->makeTransport()]);
        $list = $client->paymentMethods->list(['method_type' => 'bank_account', 'verified' => 'true']);

        self::assertInstanceOf(PaymentMethodList::class, $list);
        self::assertSame('pm-1', $list->items[0]->id);
        self::assertStringContainsString('method_type=bank_account', $this->lastUrl());
        self::assertStringContainsString('verified=true', $this->lastUrl());
    }

    public function testGet(): void
    {
        $this->enqueueOk(self::fixture());
        $client = new Client(['transport' => $this->makeTransport()]);
        $pm = $client->paymentMethods->get('pm-1');

        self::assertInstanceOf(PaymentMethod::class, $pm);
        self::assertTrue($pm->verified);
        self::assertSame('Vendor', $pm->counterparty?->name);
    }

    public function testGetMapsNotFound(): void
    {
        $this->enqueueError(404, ['error' => 'gone', 'error_code' => 'PAYMENT_METHOD_NOT_FOUND']);
        $client = new Client(['transport' => $this->makeTransport()]);
        $this->expectException(PaymentMethodNotFoundException::class);
        $client->paymentMethods->get('missing');
    }

    public function testCreateWrapsBody(): void
    {
        $this->enqueueOk(self::fixture(), 201);
        $client = new Client(['transport' => $this->makeTransport()]);
        $pm = $client->paymentMethods->create([
            'method_type' => 'bank_account',
            'currency' => 'VES',
            'label' => 'Main',
            'details' => ['bank_code' => '0102', 'account_number' => '...', 'holder_name' => 'ACME'],
        ], 'idem-pm-1');

        self::assertInstanceOf(PaymentMethod::class, $pm);
        $body = json_decode($this->lastBody() ?? '', true);
        self::assertSame('bank_account', $body['payment_method']['method_type']);
        self::assertSame('idem-pm-1', $this->lastHeaders()['Idempotency-Key']);
    }

    public function testUpdatePatchesPartial(): void
    {
        $this->enqueueOk(self::fixture());
        $client = new Client(['transport' => $this->makeTransport()]);
        $client->paymentMethods->update('pm-1', ['label' => 'Renamed']);

        self::assertSame('PATCH', $this->lastMethod());
        $body = json_decode($this->lastBody() ?? '', true);
        self::assertSame('Renamed', $body['payment_method']['label']);
    }

    public function testUpdateMapsValidation(): void
    {
        $this->enqueueError(400, ['error' => 'bad', 'error_code' => 'VALIDATION_ERROR']);
        $client = new Client(['transport' => $this->makeTransport()]);
        $this->expectException(ValidationException::class);
        $client->paymentMethods->update('pm-1', ['label' => '']);
    }

    public function testDeleteReturnsVoidOn204(): void
    {
        $this->curl->enqueue(new CurlResult(204, '', ['x-request-id' => 'req-x'], 0, ''));
        $client = new Client(['transport' => $this->makeTransport()]);
        $client->paymentMethods->delete('pm-1', 'idem-del-1');

        self::assertSame('DELETE', $this->lastMethod());
        self::assertSame('idem-del-1', $this->lastHeaders()['Idempotency-Key']);
    }

    public function testDeleteMapsValidationWhenInUse(): void
    {
        $this->enqueueError(409, ['error' => 'in use', 'error_code' => 'VALIDATION_ERROR']);
        $client = new Client(['transport' => $this->makeTransport()]);
        $this->expectException(ValidationException::class);
        $client->paymentMethods->delete('pm-1');
    }
}
